### ISSUE TITLE
Fable 3.7.1 breaks project 

### ISSUE THREAD

**Issue by forki** [fable-compiler/Fable#2826]
After update to 3.7.1 we see:

```
JavaScript error: https://gateway.dimoco-payments.eu/js/integrated/sejaqlhDfI3D1HvvHvTz.v1.3.js?_=1645091151405, line 172: TypeError: master_frame.contentWindow is null
JavaScript error: https://gateway.dimoco-payments.eu/js/jquery.creditCardValidator.js?_=1645091151408, line 33: TypeError: $ is undefined
JavaScript error: https://gateway.dimoco-payments.eu/js/integrated/sejaqlhDfI3D1HvvHvTz.v1.3.js?_=1645091151406, line 172: TypeError: master_frame.contentWindow is null
```

Downgrade to 3.7.0 seems to fix it.

**Comment by forki** [comment]
@alfonsogarciacaro  is this related to the Fulma issue that you fixed?

**Comment by forki** [comment]
maybe it's something that's already broken and fable now find's it. I'll investigate that first

**Comment by alfonsogarciacaro** [comment]
The Fulma issue only affected Fable.Core so probably not. 3.7.1 was released to fix #2809 and #2817. Maybe one of those caused the regression. It looks like code from dimoco-payments is being activated too early (or when it shouldn't be activated at all). Would it be possible to see the F# code that invokes dimoco-payments? If it cannot b shared maybe you can send it to me in private?

**Comment by forki** [comment]
In index.html we have:


        <script data-main="payment-js" src="https://gateway.dimoco-payments.eu/js/integrated/payment.1.3.min.js"></script>

and then we use it like this:

    module Client.Components.CreditCardForm

    open Elmish
    open Fable.React
    open Fable.React.Props
    open Fable.Core.JsInterop
    open Fable.I18Next
    open Fulma
    open Client.Style
    open Client.Validation
    open Fable.Dimoco
    open ServerCode.Domain.DIMOCO
    open ServerCode.Domain

    type Props = {
        PublicIntegrationKey : string
        Translate : string -> string
        OnLoaded : string -> unit
        OnChanged : CreditCardData -> unit
        Data : CreditCardData
        ValidationErrors : ValidationError []
        ShowEMail : bool
    }

    let tokenizeCreditCard tokenizer data successMsg errorMsg =
        let convert (i:CreditCardDataToken) =
            {
                Token = i.Token
                CardType = CreditCardType.Parse i.CardData?card_type
                FullName = i.CardData?full_name
                FirstSixDigits = i.CardData?first_six_digits
                LastFourDigits = i.CardData?last_four_digits
                Month = i.CardData?month
                Year = i.CardData?year
                Fingerprint = i.CardData?fingerprint
            }

        match data.ExpirationDate with
        | Some date ->
            let year =
                if date.Year < 100 then
                    (date.Year + 2000).ToString()
                else
                    date.Year.ToString()
            Cmd.OfPromise.either tokenizer.Tokenize (data.Holder,date.Month.ToString(),year,data.EMailAddress) (convert >> successMsg) errorMsg
        | None ->
            Cmd.OfPromise.either tokenizer.Tokenize (data.Holder,"","",data.EMailAddress) (convert >> successMsg) errorMsg

    let renderSupportedCreditCardTypes translate =
        p [] [
            str (I18n.Translate "FAQ_Entries.SupportedCreditCards.Paragraph1")
            let supportedTypes =
                CreditCardType.All
                |> Array.filter (fun t -> t.IsSupported())
                |> Array.map (fun t -> t.Translate())

            str (System.String.Join (", ", supportedTypes))
        ]

    let private mapErrorMessages attribute key =
        match attribute with
        | "card_holder" ->
            match key with
            | "errors.blank" -> "Card holder must not be empty."
            | _ -> ""
        | "number" ->
            match key with
            | "errors.blank" -> "Card number must not be empty."
            | "errors.invalid" -> "Invalid card number."
            | "errors.unsupported" -> "Card type is not supported."
            | _ -> ""
        | "cvv" ->
            match key with
            | "errors.blank" -> "CVV code must not be empty."
            | "errors.invalid" -> "Invalid CVV code."
            | _ -> ""
        | "month"
        | "year" ->
            match key with
            | "errors.blank" -> "Expiration date must not be empty."
            | "errors.invalid" -> "Invalid expiration date."
            | "errors.expired" -> "Card expired."
            | _ -> ""
        | _ -> ""

    let mapValidationErrors errors =
        let tokenErrors : CreditCardTokenError [] = unbox errors // actually it's an error []
        let validationErrors =
            tokenErrors
            |> Array.map (fun e -> {
                Key = e.Attribute
                Message = mapErrorMessages e.Attribute e.Key })
        validationErrors

    let CreditCardForm = FunctionComponent.Of(fun (props:Props) ->
        let data = Hooks.useState(props.Data)

        let onValueChanged modifier value =
            data.update
                (fun state ->
                    let state = modifier state value
                    props.OnChanged state
                    state)

        let register() =
            props.OnLoaded props.PublicIntegrationKey

        Hooks.useEffect(register,[| props.PublicIntegrationKey |])

        let getError key =
            match getErrorForKey props.ValidationErrors key with
            | Some error ->
                Some (props.Translate error.Message)
            | None -> None

        form [] [
            renderSupportedCreditCardTypes props.Translate

            p [ Style [ Color "red"] ] [
                str (props.Translate "VISA payment is temporarily suspended. Please use Mastercard or PayPal. We are sorry for the inconvenience.")
            ]

            horizontalFieldRequired "cardholder" (props.Translate "Card holder") (onValueChanged (fun state value -> { state with Holder = value })) (getError "card_holder") data.current.Holder

            horizontalFieldRequiredContent "cardnumber" (props.Translate "Card number") (getError "number") (
                div [ ClassName "input"; Id "number_div" ] []
            )

            horizontalFieldRequiredContent "cvv" (props.Translate "CVV") (getError "cvv") (
                div [ ClassName "input"; Id "cvv_div" ] []
            )

            let dateError =
                getError "month"
                |> Option.orElse (getError "year")

            horizontalFieldRequiredContent "expirationDate" (props.Translate "Expiration date") dateError (
                Patterns.Form.CustomFormatDateTimeInput ({
                    Label = ""
                    Value = Option.map (fun value -> Formatting.localToUTC (System.DateTimeOffset value)) data.current.ExpirationDate
                    OnChange = (fun date ->
                        let value = date |> Option.map (fun value -> (Formatting.utcToLocal value).DateTime)

                        onValueChanged (fun state value -> { state with ExpirationDate = value }) value
                    )
                    Optional = []
                }, "m/Y")
            )

            if props.ShowEMail then
                horizontalFieldRequiredContent "email" (props.Translate "E-Mail") (getError "email") (
                    Input.email [
                        Input.OnChange (fun ev -> onValueChanged (fun state value -> { state with EMailAddress = value }) ev.Value)
                        Input.Value data.current.EMailAddress
                    ]
                )

            requiredHint ()
        ]
    )

tbh I'd like to get rid of the line in index.html but did not get that working.

**Comment by alfonsogarciacaro** [comment]
Thanks for sharing @forki. Does the Fable.Dimoco module contain Fable bindings for Dimoco? I assume they're private as I cannot find them in Github nor Nuget. Could these be shared as well? It'd also be helpful to see how the code causing the errors is activated, so having the stack trace would be great too :)

**Comment by alfonsogarciacaro** [comment]
@forki Scratch that, reviewing the code I think I made a mistake when fixing #2817. Can you please try with Fable v3.7.2 and check if it works now?

**Comment by forki** [comment]
module Fable.Dimoco

    open Fable.Core
    open Fable.Core.JsInterop

    type CreditCardDataToken = {
        Token : string
        CardData : obj
    }

    let InitPaymentForm (publicIntegrationKey:string) : JS.Promise<obj> = importMember "./js/dimoco.js"
    let Tokenize(cardHolder:string,month:string,year:string,email:string) : JS.Promise<CreditCardDataToken> = importMember "./js/dimoco.js"

    type Tokenization = {
        Tokenize : string * string * string * string -> JS.Promise<CreditCardDataToken>
    }
        with
            static member Default = {
                Tokenize = Tokenize
            }

**Comment by forki** [comment]
const payment = new PaymentJs();

    export function InitPaymentForm(publicIntegrationKey) {
        return new Promise(resolve => {
            payment.init(publicIntegrationKey, 'number_div', 'cvv_div', payment => {
                const style = {
                    'border': '0',
                    'outline': '0'
                };

                payment.setNumberStyle(style);
                payment.setCvvStyle(style);

                resolve(payment);
            });
        });
    }

    export function Tokenize(cardHolder, month, year, email) {
        const cardTypeSupported = (cardType) =>
            (cardType === 'visa' || cardType === 'visa_electron' || cardType === 'mastercard' || cardType === 'maestro');

        return new Promise((resolve, reject) => {
            var data = {
                card_holder: cardHolder,
                month,
                year,
                email
            };

            payment.tokenize(
                data,
                (token, cardData) => {
                    if (cardTypeSupported(cardData.card_type)) {
                        resolve({
                            Token: token,
                            CardData: cardData
                        });
                    } else {
                        reject([
                            {
                                Attribute: 'number',
                                Key: 'errors.unsupported',
                                Message: 'Card type is not supported',
                            },
                        ]);
                    }
                },
                errors => {
                    reject(errors.map(error => ({
                        Attribute: error.attribute,
                        Key: error.key,
                        Message: error.message,
                    })));
                }
            );
        });
    }

**Comment by forki** [comment]
the first one ist the bindings. The second one is our own dimoco.js

Hope that helps. Sorry that I forgot that before

**Comment by forki** [comment]
with Fable 3.7.2 we get:

```
JavaScript warning: https://gateway.dimoco-payments.eu/js/integrated/payment.1.3.min.js, line 27: WebGL warning: <Create>: WebglAllowWindowsNativeGl:false restricts context creation on this system.
JavaScript warning: https://gateway.dimoco-payments.eu/js/integrated/payment.1.3.min.js, line 27: Failed to create WebGL context: WebGL creation failed: 
* WebglAllowWindowsNativeGl:false restricts context creation on this system. ()
* Exhausted GL driver options. (FEATURE_FAILURE_WEBGL_EXHAUSTED_DRIVERS)
JavaScript warning: https://gateway.dimoco-payments.eu/js/integrated/payment.1.3.min.js, line 27: WebGL warning: <Create>: WebglAllowWindowsNativeGl:false restricts context creation on this system.
JavaScript warning: https://gateway.dimoco-payments.eu/js/integrated/payment.1.3.min.js, line 27: Failed to create WebGL context: WebGL creation failed: 
* WebglAllowWindowsNativeGl:false restricts context creation on this system. ()
* Exhausted GL driver options. (FEATURE_FAILURE_WEBGL_EXHAUSTED_DRIVERS)
JavaScript warning: https://gateway.dimoco-payments.eu/js/integrated/payment.1.3.min.js, line 27: WebGL warning: <Create>: WebglAllowWindowsNativeGl:false restricts context creation on this system.
JavaScript warning: https://gateway.dimoco-payments.eu/js/integrated/payment.1.3.min.js, line 27: Failed to create WebGL context: WebGL creation failed: 
* WebglAllowWindowsNativeGl:false restricts context creation on this system. ()
* Exhausted GL driver options. (FEATURE_FAILURE_WEBGL_EXHAUSTED_DRIVERS)
JavaScript warning: https://gateway.dimoco-payments.eu/js/integrated/payment.1.3.min.js, line 27: WebGL warning: <Create>: WebglAllowWindowsNativeGl:false restricts context creation on this system.
JavaScript warning: https://gateway.dimoco-payments.eu/js/integrated/payment.1.3.min.js, line 27: Failed to create WebGL context: WebGL creation failed: 
* WebglAllowWindowsNativeGl:false restricts context creation on this system. ()
* Exhausted GL driver options. (FEATURE_FAILURE_WEBGL_EXHAUSTED_DRIVERS)
```

and later:

```
JavaScript error: https://gateway.dimoco-payments.eu/js/integrated/sejaqlhDfI3D1HvvHvTz.v1.3.js?_=1645512598631, line 172: TypeError: master_frame.contentWindow is null
JavaScript error: https://gateway.dimoco-payments.eu/js/integrated/sejaqlhDfI3D1HvvHvTz.v1.3.js?_=1645512598630, line 172: TypeError: master_frame.contentWindow is null
```

**Comment by forki** [comment]
reverting to 3.7.0 fixed it again

**Comment by forki** [comment]
or rather it changes it to 

```
JavaScript error: https://gateway.dimoco-payments.eu/js/integrated/sejaqlhDfI3D1HvvHvTz.v1.3.js?_=1645513341353, line 427: TypeError: this.masterFrame is null
JavaScript error: https://gateway.dimoco-payments.eu/js/jquery.creditCardValidator.js?_=1645513341355, line 33: TypeError: $ is undefined
```

which makes the tests pass

**Comment by alfonsogarciacaro** [comment]
Not sure why reverting Fable causes different behavior but looking at your code for dimoco.js, I see that you initialize `payment` with 4 arguments:

```js
payment.init(publicIntegrationKey, 'number_div', 'cvv_div', payment => {
```

However when I check the code for `https://gateway.dimoco-payments.eu/js/integrated/sejaqlhDfI3D1HvvHvTz.v1.3.js` the init function only has 3 arguments, so I assume that `numberDivId` gets the incorrect value so the element is not located on the page and that's why the "master_frame.contentWindow is null" error is raised.

![image](https://user-images.githubusercontent.com/8275461/155083105-c976c725-b3b7-49c3-96e9-aad006e0b78f.png)

**Comment by alfonsogarciacaro** [comment]
Hmm, maybe It's not exactly the same `init` function being invoked by your code. In any case, can you please check that there's actually an element in your page with id "number_div"?

**Comment by forki** [comment]
thanks for making me double check this. I think I found one issue.

The problem is that we actually have two instances of the credit card form on the site. One for mobile view and one for desktop. Those divs have different CSS classes so that browsers will only show one at the time (depending on screen width). Now I totally forgot that both forms are still in the DOM and actually compete over the `number_div` ID. After we resolved this competition then everything works nicely without any error message in Fable 3.7.0.

Well almost. There is one small issue left:

    let register() =
        props.OnLoaded props.PublicIntegrationKey
    Hooks.useEffect(register,[| props.PublicIntegrationKey |])

in earlier versions we used 

    override __.componentDidMount() =
        props.OnLoaded props.PublicIntegrationKey

I think it would be more correct to use `Hooks.useLayoutEffect`  to actually run it after the first render. But `Hooks.useLayoutEffect` doesn't exist.

Also: It still fails with Fable 3.7.2 - BUT this I think this is unrelated to this issue here. So I will open a new one. 
Only thing left here is the question if Hooks should get a useLayoutEffect

**Comment by alfonsogarciacaro** [comment]
Yes, it should be fine to add `useLayoutEffect` to Fable.React. Can you please send a PR? As a temporary solution you can use an extension:

```fsharp
type IHooks with
  [<Emit("$0.useLayoutEffect($1...)")>]
  member _.useLayoutEffect(effect: unit->unit, ?dependencies: obj[]) = ()
```

This won't do anything in SSR but in JS code Fable will replace it with the contents of Emit.

**Comment by alfonsogarciacaro** [comment]
New try :) I've pushed 3.7.3 which should fix #2832. Can you please try upgrading and close this issue if it finally works for you? 🤞

**Comment by forki** [comment]
Thanks so much for helping me. Everything is green now!!
